google/cloud/bigtable/_row_merger.py

```diff
@@ -0,0 +1,621 @@
+            )
+        new_gapic_stream = await gapic_fn(self.request, timeout=per_request_timeout)
+        cache: asyncio.Queue[Row | RequestStats] = asyncio.Queue(maxsize=cache_size)
+        state_machine = _StateMachine()
+        try:
+            stream_task = asyncio.create_task(
+                _RowMerger._generator_to_cache(
+                    cache,
+                    _RowMerger.merge_row_response_stream(
+                        new_gapic_stream, state_machine
+                    ),
+                )
+            )
+            get_from_cache_task = asyncio.create_task(cache.get())
+            # sleep to allow other tasks to run
+            await asyncio.sleep(0)
+            # read from state machine and push into cache
+            # when finished, stream will be done, cache will be empty, but get_from_cache_task will still be waiting
+            while (
+                not stream_task.done()
+                or not cache.empty()
+                or get_from_cache_task.done()
+            ):
+                if get_from_cache_task.done():
+                    new_item = get_from_cache_task.result()
+                    # don't yield rows that have already been emitted
+                    if isinstance(new_item, RequestStats):
+                        yield new_item
+                    elif (
+                        isinstance(new_item, Row)
+                        and new_item.row_key not in self.emitted_rows
+                    ):
+                        self.last_seen_row_key = new_item.row_key
+                        # don't yeild _LastScannedRow markers; they
+                        # should only update last_seen_row_key
+                        if not isinstance(new_item, _LastScannedRow):
+                            self.emitted_rows.add(new_item.row_key)
+                            yield new_item
+                    # start new task for cache
+                    get_from_cache_task = asyncio.create_task(cache.get())
+                    await asyncio.sleep(0)
+                else:
+                    # wait for either the stream to finish, or a new item to enter the cache
+                    first_finish = asyncio.wait(
+                        [stream_task, get_from_cache_task],
+                        return_when=asyncio.FIRST_COMPLETED,
+                    )
+                    await asyncio.wait_for(first_finish, per_row_timeout)
+            # stream and cache are complete. if there's an exception, raise it
+            if stream_task.exception():
+                raise cast(Exception, stream_task.exception())
+        except asyncio.TimeoutError:
+            # per_row_timeout from asyncio.wait_for
+            raise core_exceptions.DeadlineExceeded(
+                f"per_row_timeout of {per_row_timeout:0.1f}s exceeded"
+            )
+        finally:
+            stream_task.cancel()
+            get_from_cache_task.cancel()
+
+    @staticmethod
+    def _revise_request_rowset(
```

Comment:
We should also check the limit of the rows. If it's set, we should update it to (originalLimit - processedRows). If the number of returned rows == limit, don't trigger a retry.

Also add a test where the original request has a limit to test this behavior.

Reply:
Good point, I made the change, but will leave this comment up until I add tests